src/nbody_base.jl

```diff
@@ -0,0 +1,157 @@
+using StaticArrays, DiffEqBase
+#Structures for bodies and systems under symulations
+
+# Fields for position, velocity and mass of a particle should be required for every descendant
+abstract type Body
+
+end
+
+struct SphericalBody{mType, qType, cType <: AbstractFloat} <: Body
+    r :: SVector{3, cType}  #required 
+    v :: SVector{3, cType}  #required
+    m :: mType              #required
+    R :: cType
+    q :: qType 
+    mm :: SVector{3, qType}
+end
+
+abstract type NBodySystem
+end
+
+struct MDSystem{bType<:Body} <: NBodySystem
+    bodies :: Array{bType, 1}
+    thermostats :: Array{Symbol, 1}
+    collisions :: Array{Symbol, 1}
+end
+
+struct ChargedParticles{bType<:Body} <: NBodySystem
+    bodies :: Array{bType, 1}
+    k :: AbstractFloat
+end
+
+struct GravitationalInteractions{bType<:Body} <: NBodySystem
+    bodies :: Array{bType, 1}
+end
+
+struct GenericInteractingBodies{bType<:Body} <: NBodySystem
+    bodies :: Array{bType, 1}
+    pairwise_potential
+end
+
+# It is convenient to load data for particles from a file
+function load_system_from_file(path::AbstractString)
+    f = open(path)
+    data = readlines(f)
+    n = size(data, 1)
+    data = load_file(path);
+    bodies = Body[]
+    for i=1:n
+        #body = compile_body(data[i])
+        #push!(bodies, body)
+    end
+    
+    return NBodySystem(bodies)
+end
+
+# This structure defines conditions under wich we test our sysmem of n-bodies
+struct NBodySimulation{sType, cType <: AbstractFloat}
+    system :: sType
+    limiting_boundary :: SVector{6, cType}
+    tspan :: Tuple{Float64, Float64}
+    external_electric_field
+    external_magnetic_field
+    external_gravitational_field
+end
+
+NBodySimulation(system::NBodySystem, limiting_boundary :: SVector{6, Float64}, tspan :: Tuple{Float64, Float64}) = 
+    NBodySimulation(system, limiting_boundary, tspan, x->0, x->0, x->0)
+
+# Instead of treating NBodySimulation as a DiffEq problem and passing it into a solve method
+# it is better to use a specific function for n-body simulations.
+function run_simulation(s::NBodySimulation, args...; kwargs...)
+    #create a DEProblem depending on types of interactions between particles
+    solution = solve(problem) 
+    return SimulationResult(solution)
+end
+
+# Function as this one can be used in a force-oriented approach of calculations
+function acceleration(
+    system :: NBodySystem, 
+    external_electric_field,
+    external_magnetic_field,
+    external_gravitational_field)
+end
+
+# Function, scpecifically designed for a system of ChargedParticles
```

Comment:
I don't get this part. What's going on here? Are you using different acceleration functions based on the existence of some fields and not other?

Reply:
This is unclear, I agree. No, not based on the presence or absence of external fields. 
Actually, the second implementation is wrong - I want to pass `NBodySystem` into acceleration function and there to decide what forces act on i-th particle. In addition to the forces from other particles the external fields also affect the i-th particle. But then I realized, that I need the positions of bodies. So here is a mess, I agree.
The better definition of this function would be something like this:
```julia
function acceleration(
   u,
   i,
   system :: NBodySystem, 
   external_electric_field,
   external_magnetic_field,
   external_gravitational_field)
+end
```

Reply:
Looks like passing the NBodySimulation could make sense?